common/src/main/java/org/vivecraft/client_vr/gameplay/VRPlayer.java

```diff
@@ -249,208 +338,235 @@ public void setRoomOrigin(double x, double y, double z, boolean reset) {
+                this.roomOrigin,
+                this.dh.vrSettings.walkMultiplier,
+                this.worldScale,
+                (float) Math.toRadians(this.dh.vrSettings.worldRotation));
+        }
+
+        Vec3 camPos = temp.getHeadPivot().subtract(temp.origin);
+        double x = player.getX() - camPos.x;
+        double y = player.getZ() - camPos.z;
+        double z = player.getY() + ((PlayerExtension) player).vivecraft$getRoomYOffsetFromPose();
+        this.setRoomOrigin(x, z, y, reset);
     }
 
+    // calculate the shortest difference between 2 angles.
     public float rotDiff_Degrees(float start, float end) {
-        double d0 = Math.toRadians(end);
-        double d1 = Math.toRadians(start);
-        return (float) Math.toDegrees(Math.atan2(Math.sin(d0 - d1), Math.cos(d0 - d1)));
+        double x = Math.toRadians(end);
+        double y = Math.toRadians(start);
+        return (float) Math.toDegrees(Math.atan2(Math.sin(x - y), Math.cos(x - y)));
     }
 
     public void rotateOriginAround(float degrees, Vec3 o) {
-        Vec3 vec3 = this.roomOrigin;
-        float f = (float) Math.toRadians(degrees);
+        Vec3 point = this.roomOrigin;
 
-        if (f != 0.0F) {
-            this.setRoomOrigin(Math.cos(f) * (vec3.x - o.x) - Math.sin(f) * (vec3.z - o.z) + o.x, vec3.y, Math.sin(f) * (vec3.x - o.x) + Math.cos(f) * (vec3.z - o.z) + o.z, false);
+        // reverse rotate.
+        float rads = (float) Math.toRadians(degrees);
+
+        if (rads != 0.0F) {
+            this.setRoomOrigin(
+                Math.cos(rads) * (point.x - o.x) - Math.sin(rads) * (point.z - o.z) + o.x,
+                point.y,
+                Math.sin(rads) * (point.x - o.x) + Math.cos(rads) * (point.z - o.z) + o.z,
+                false);
         }
     }
 
-    public void tick(LocalPlayer player, Minecraft mc, RandomSource rand) {
-        if (((PlayerExtension) player).vivecraft$getInitFromServer()) {
-            if (!this.initdone) {
-                System.out.println("<Debug info start>");
-                System.out.println("Room object: " + this);
-                System.out.println("Room origin: " + this.vrdata_world_pre.origin);
-                System.out.println("Hmd position room: " + this.vrdata_room_pre.hmd.getPosition());
-                System.out.println("Hmd position world: " + this.vrdata_world_pre.hmd.getPosition());
-                System.out.println("Hmd Projection Left: " + dh.vrRenderer.eyeproj[0]);
-                System.out.println("Hmd Projection Right: " + dh.vrRenderer.eyeproj[1]);
-                System.out.println("<Debug info end>");
-                this.initdone = true;
-            }
+    public void tick(LocalPlayer player, Minecraft mc) {
+        if (!((PlayerExtension) player).vivecraft$getInitFromServer()) {
+            return;
+        }
 
-            this.doPlayerMoveInRoom(player);
+        if (!this.initDone) {
+            System.out.println("<Debug info start>");
```

Comment:
this should also use the logger no?

Reply:
oops missed that, but this is still a work in prograss, ther are still many system outs I think